### PR TITLE
Split Emag in half

### PR DESCRIPTION
**Pull request by WarMechanic**
## About the PR

Splits the functionality of Emag into two different items:
- Cryptographic Sequencer (4tc), which can emag everything except doors.
- Airlock Access Override (6tc), which can only emag doors.

## Why / Balance

Emag is a very big item which makes it hard to balance. On one hand, its a gimmick enabler by creating special interactions with station machines. On the other hand, it is very efficient at breaking and entering anywhere with a door. Most players will want to buy the Emag for one function or the other.

By splitting Emag into these two functions, the collective cost of Emag will stay at 10 but Syndicate players will have more agency over what they want to do in a round with budget cuts. The cost of either item can be interpolated based on which one is more powerful.

## Technical details

Adds whitelist and blacklist variables to the Emag component.

## Media

https://youtu.be/c0NuEvLdc_U
image
image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

No

**Changelog**

:cl:
- add: Split the Emag into the Cryptographic Sequencer and the Airlock Access Override.

### CODE REVIEW

**Comment by deltanedas** [comment]
> This PR is currently awaiting a sprite for the AAO.

doesnt tg have it

**Comment by WarMechanic** [comment]
> > This PR is currently awaiting a sprite for the AAO.
> 
> doesnt tg have it

it does but it also has different cryptographic sequencer and id card sprites which would look out of place here. i guess i could use it temporarily for testing but i dont know how attributions work so i dont want to bother lol.
i got someone on discord to work on a new sprite so we good.

File: Content.Shared/Emag/Systems/EmagSystem.cs [hunk 1]
```diff
@@ -50,6 +52,15 @@ public bool TryUseEmag(EntityUid uid, EntityUid user, EntityUid target, EmagComp
         if (_tag.HasTag(target, comp.EmagImmuneTag))
             return false;
 
+        // DeltaV - Add a whitelist / blacklist to the Emag
+        if (_whitelistSystem.IsWhitelistFail(comp.Whitelist, target)
+            || _whitelistSystem.IsWhitelistPass(comp.Blacklist, target))
+        {
+            _popup.PopupClient("Invalid target!", user, user); // do i really need to use a locale for this?
```
Comment (Mnemotechnician) [inline Content.Shared/Emag/Systems/EmagSystem.cs]: I suggest adding a locale string `emag-invalid-target = {$emag} has no effect on ${target}!` and replacing this with `_popup.PopupClient(Loc.GetString("emag-invalid-target", ("emag", uid), ("target", target)), user, user);`.

Comment (WarMechanic) [reply Content.Shared/Emag/Systems/EmagSystem.cs]: done :)

**Comment by Lyndomen** [comment]
I think the admin team is really excited about this, and look forward to it if it is implemented.

File: Resources/Locale/en-US/deltav/store/uplink-catalog.ftl [hunk 2]
```diff
@@ -13,3 +13,6 @@ uplink-syndicate-radio-implanter-bundle-desc = Two implanters for the price of o
 
 uplink-generic-radio-implanter-name = Generic Radio Implanter
 uplink-generic-radio-implanter-desc = A cranial implant with a bluespace compartment for a single encryption key (not included). Put in a key of your choice, and you can talk using it like you would with any headset.
+
+uplink-doorjack-name = Airlock Access Override
+uplink-doorjack-desc = A specialized cryptographic sequencer, designed solely to doorjack Nanotrasen's updated airlocks. Does not tamper with anything else.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/deltav/store/uplink-catalog.ftl]:
```suggestion
uplink-doorjack-desc = A specialized cryptographic sequencer, designed solely to doorjack NanoTrasen's updated airlocks. Does not tamper with anything else.

```

File: Resources/Locale/en-US/emag/emag.ftl [hunk 3]
```diff
@@ -1,2 +1,4 @@
 emag-success = The card zaps something in {THE($target)}.
 emag-no-charges = No charges left!
+# DeltaV
+emag-invalid-target = {$emag} has no effect on {THE($target)}.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/emag/emag.ftl]:
```suggestion
emag-invalid-target = {$emag} has no effect on {THE($target)}.

```

File: Resources/Locale/en-US/store/uplink-catalog.ftl [hunk 4]
```diff
@@ -112,8 +112,9 @@ uplink-holster-desc = A deep shoulder holster capable of holding many types of b
 uplink-chest-rig-name = Chest Rig
 uplink-chest-rig-desc = Explosion-resistant tactical webbing used for holding traitor goods.
 
-uplink-emag-name = Emag
-uplink-emag-desc = The business card of the syndicate, this sequencer is able to break open airlocks and tamper with a variety of station devices. Recharges automatically.
+# DeltaV - Correct Emag name, Emag no longer breaks doors
+uplink-emag-name = Cryptographic Sequencer 
+uplink-emag-desc = The business card of the syndicate, this sequencer is able to tamper with a variety of station devices. Rendered obsolete against airlocks.
```
Comment (DEATHB4DEFEAT) [inline Resources/Locale/en-US/store/uplink-catalog.ftl]:
```suggestion
uplink-emag-desc = The business card of the Syndicate, this sequencer is able to tamper with a variety of station devices. Rendered obsolete against airlocks.
```

File: Resources/Prototypes/DeltaV/Catalog/uplink_catalog.yml [hunk 5]
```diff
@@ -71,3 +71,13 @@
...
+
+- type: listing
+  id: UplinkDoorjack
+  name: uplink-doorjack-name
+  description: uplink-doorjack-desc
+  productEntity: Doorjack
+  cost:
+    Telecrystal: 6
+  categories:
+  - UplinkDisruption
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/DeltaV/Catalog/uplink_catalog.yml]:
```suggestion
  - UplinkDisruption

```

File: Resources/Prototypes/DeltaV/Entities/Objects/Tools/emag.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    blacklist:
+    whitelist:
+      components:
+      - Airlock
+  - type: Sprite
+    sprite: DeltaV/Objects/Tools/doorjack.rsi
+    state: icon
+  - type: Item
+    sprite: DeltaV/Objects/Tools/doorjack.rsi
+    storedRotation: -90
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/DeltaV/Entities/Objects/Tools/emag.yml]:
```suggestion
    storedRotation: -90

```

**Review by DEATHB4DEFEAT — DISMISSED** [review]

File: Content.Shared/Emag/Components/EmagComponent.cs [hunk 7]
```diff
@@ -17,4 +18,18 @@ public sealed partial class EmagComponent : Component
...
+    /// Whitelist that entities must be on to work.
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite), AutoNetworkedField]
+    public EntityWhitelist? Whitelist = null;
+
+    /// <summary>
+    /// Blacklist that entities must be off to work.
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite), AutoNetworkedField]
+    public EntityWhitelist? Blacklist = null;
```
Comment (deltanedas) [inline Content.Shared/Emag/Components/EmagComponent.cs]: dont need ViewVariables, AutoNetworkedField or = null:
- datafield is automatically rw
- nothing ever changes them so they never get dirtied so they never get sent
- null is the default

File: Content.Shared/Emag/Systems/EmagSystem.cs [hunk 8]
```diff
@@ -23,6 +24,7 @@ public sealed class EmagSystem : EntitySystem
     [Dependency] private readonly SharedChargesSystem _charges = default!;
     [Dependency] private readonly SharedPopupSystem _popup = default!;
     [Dependency] private readonly TagSystem _tag = default!;
+    [Dependency] private readonly EntityWhitelistSystem _whitelistSystem = default!; // DeltaV - Add a whitelist/blacklist to the Emag
```
Comment (deltanedas) [inline Content.Shared/Emag/Systems/EmagSystem.cs]: rename to _whitelist

File: Content.Shared/Emag/Systems/EmagSystem.cs [hunk 9]
```diff
@@ -50,6 +52,15 @@ public bool TryUseEmag(EntityUid uid, EntityUid user, EntityUid target, EmagComp
         if (_tag.HasTag(target, comp.EmagImmuneTag))
             return false;
 
+        // DeltaV - Add a whitelist / blacklist to the Emag
+        if (_whitelistSystem.IsWhitelistFail(comp.Whitelist, target)
+            || _whitelistSystem.IsWhitelistPass(comp.Blacklist, target))
```
Comment (deltanedas) [inline Content.Shared/Emag/Systems/EmagSystem.cs]: IsBlacklistPass for clarity

**Review by deltanedas — CHANGES_REQUESTED** [review]

**Comment by WarMechanic** [comment]
all requested changes are resolved, all checks have passed and it still works in game so it looks good to me

**Comment by WarMechanic** [comment]
only a week away

**Comment by Axionyxx** [comment]
@deltanedas @DEATHB4DEFEAT Have your things been properly addressed?

**Review by deltanedas — APPROVED** [review]

**Review by Axionyxx — APPROVED** [review]
Only have a few small nitpicks, but they don't matter enough for me to actually type them out and click the request changes button

**Pull request merged**